src/super_gradients/module_interfaces/supports_input_shape_check.py

```diff
@@ -0,0 +1,60 @@
+from typing import Tuple, Optional
+
+
+class SupportsInputShapeCheck:
+    def validate_input_shape(self, input_shape: Tuple[int, ...]) -> None:
+        """
+        Validate the input shape of the model.
+        If the validation passes, the method should return None.
+        If the validation fails, the method should raise a ValueError.
+
+        :param input_shape: (Tuple[int, ...]) Input shape (usually BCHW)
+        :return: None.
+        """
+        exact_shape = self.get_exact_input_shape_size()
+        if exact_shape is not None:
+            # Take n last elements of input shape, where n is the number of elements in exact_shape
+            shape_to_check = input_shape[-len(exact_shape) :]
+            if any(shape_to_check != exact_shape):
+                raise ValueError(f"Input shape {input_shape} is not supported. Model requires input with spatial size of {exact_shape}.")
+
+        min_shape = self.get_minimum_input_shape_size()
+        if min_shape is not None:
+            shape_to_check = input_shape[-len(min_shape) :]
+            if any(shape_to_check < min_shape):
+                raise ValueError(f"Input shape {input_shape} is not supported. Model requires input with spatial size of at least {min_shape}.")
+
+        shape_steps = self.get_input_shape_steps()
+        if shape_steps is not None:
+            shape_to_check = input_shape[-len(shape_steps) :]
+            if any(shape_to_check % shape_steps != 0):
+                raise ValueError(f"Input shape {input_shape} is not supported. Model requires input with spatial size that is a multiple of {shape_steps}.")
+
+    def get_exact_input_shape_size(self) -> Optional[Tuple[int, ...]]:
+        """
+        Get the exact input shape of the model.
+        Class should override this method if the model requires an exact input shape.
+        Usually true for transformer-based models.
+        :return:
+        """
+        return None
+
+    def get_minimum_input_shape_size(self) -> Optional[Tuple[int, ...]]:
+        """
+        Get the minimum input shape of the model.
+        Class should override this method if the model requires a minimum input shape.
+        Usually true for transformer-based models.
+        :return:
+        """
+        return None
+
+    def get_input_shape_steps(self) -> Optional[Tuple[int, ...]]:
```

Comment:
The name of the method for sure can be improved. I had hard time formulating it's name )

Reply:
> Adding a new mixin is adding new code to each class that wants to inherit it. Here for instance, we needs to add 3 method to the models, on top of all the methods already required to these models. 

The idea of this mixin was to have the clear interface. The default implementation does not require one to implement any of these methods. So one can derive `SgModule(..., SupportsInputShapeCheck) and it would not break existing code. However down the class hierarcy, one can implement some methods to validate input shape. 

> I believe a proper way of scaling features would be more through composition with classes responsible for these features, not inheritance like we do with all the mixins.

Can you please show an example in pseud-code what you mean? I'm not sure I'm getting it.

Reply:
> The idea of this mixin was to have the clear interface. The default implementation does not require one to implement any of these methods. So one can derive `SgModule(..., SupportsInputShapeCheck) and it would not break existing code. However down the class hierarcy, one can implement some methods to validate input shape.

Sure I'm aware of that, and I believe it's overall a great way of doing it. I was just raising the downsides, which are that it eventually leads to very "flat" classes with lower cohesion + when someone wants to implement a new model, he needs to remember all the mixin that is available if he wants to support these features.


> Can you please show an example in pseud-code what you mean? I'm not sure I'm getting it.

Well generally speaking I don't know, and that's the problem 😅 

In this specific case, I see something we could maybe improve but it would not remove the need for mixins.

``` python
class InputValidator:
    def __init__(self, exact_image_size, min_image_size, multiple_shape):
        ...

    def validate(self, image):
        ...


class SupportsInputShapeCheck(ABC):
    @staticmethod
    @abstractmethod
    def validate_image(image):
        ...


class MyModel(SupportsInputShapeCheck):
    @staticmethod
    def validate_image(image):
        return InputValidator(exact_image_size=..., min_image_size=..., multiple_shape=...).validate(image)
        # return validate_image(exact_image_size=..., min_image_size=..., multiple_shape=..., image=image) # OR as a function directly

MyModel.validate_image(image)
```

Basically, the logic is moved outside of the Model.

But as you can see, this would still require the Mixin to declare which function support this feature.
So it does reduce the amount of methods to implement per class, and IMO leads to lower code complexity, but it does not remove the need for the inheritance of mixin.